File: nav/rcn/src/NMGen.cpp
/*
 * Copyright (c) 2010 Stephen A. Pratt
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
#include <math.h>
#include <string.h>
#include "RCN.h"
#include "RecastAlloc.h"
#include "DetourNavMeshBuilder.h"

namespace org 
{
namespace critterai
{
namespace nav
{
namespace rcn
{
    int getSolidSpanCount(const rcHeightfield& solid)
    {
	    const int w = solid.width;
	    const int h = solid.height;
        int result = 0;
    	
	    for (int y = 0; y < h; ++y)
	    {
		    for (int x = 0; x < w; ++x)
		    {
			    for (rcSpan* s = solid.spans[x + y*w]; s; s = s->next)
			    {
                    result++;
                }
            }
        }
        return result;
    }

    int getTraversableSpanCount(const rcHeightfield& solid)
    {
	    const int w = solid.width;
	    const int h = solid.height;
        int result = 0;
    	
	    for (int y = 0; y < h; ++y)
	    {
		    for (int x = 0; x < w; ++x)
		    {
			    for (rcSpan* s = solid.spans[x + y*w]; s; s = s->next)
			    {
                    result += (s->area == RC_NULL_AREA) ? 0 : 1;
                }
            }
        }
        return result;
    }

    int getTraversableSpanCount(const rcCompactHeightfield& chf)
    {
	    const int w = chf.width;
	    const int h = chf.height;
        int result = 0;
	    for (int y = 0; y < h; ++y)
	    {
		    for (int x = 0; x < w; ++x)
		    {
			    const rcCompactCell& c = chf.cells[x+y*w];
			    for (int i = (int)c.index, ni = (int)(c.index+c.count)
                    ; i < ni
                    ; i++)
			    {
				    if (chf.areas[i] != RC_NULL_AREA)
					    result++;
                }
            }
        }
        return result;
    }

    void flattenDetailMesh(const rcPolyMeshDetail& detailMesh, Mesh3& resultMesh)
    {
        /*
         * Remember: The tris array has a stride of 4: 3 indices + flags
         *
         * The detail meshes are completely independant, which results
         * in duplicate verts.  This process removes the duplicates.
         */ 

        resultMesh.vertsPerPolygon = 3;
        resultMesh.indices = new int[(detailMesh.ntris)*3];
        // Can't initialize the vertices yet.  Don't know the unique vert count.
        
        float* uniqueVerts = new float[(detailMesh.nverts)*3];
        int* vertMap = new int[detailMesh.nverts];

        int resultVertCount = removeDuplicateVerts(
            detailMesh.nverts
            , detailMesh.verts
            , uniqueVerts
            , vertMap);

        int resultVertLength = resultVertCount*3;
        float* vertices = new float[resultVertLength];

        // Copy unique vertices to the output.
        // Using resultVertLength avoids copying the trash since uniqueVerts
        // was sized to assume no duplicate vertices. (An unlikely scenario.)
        for (int i = 0; i < resultVertLength; i++)
        {
            vertices[i] = uniqueVerts[i];
        }

        // Assign vert data to the result mesh.
        resultMesh.vertices = vertices;
        resultMesh.vertCount = resultVertCount;

        delete [] uniqueVerts;

        // Flatten and re-map the indices.
        int pCurrentTri = 0;
        for (int iMesh = 0; iMesh < detailMesh.nmeshes; iMesh++)
        {
            int vBase = detailMesh.meshes[iMesh*4+0];
            int tBase = detailMesh.meshes[iMesh*4+2];
            int tCount =  detailMesh.meshes[iMesh*4+3];
            for (int iTri = 0; iTri < tCount; iTri++)
            {
                const unsigned char* tri = &detailMesh.tris[(tBase+iTri)*4];
                for (int i = 0; i < 3; i++)
                {
                    resultMesh.indices[pCurrentTri] = vertMap[vBase+tri[i]];
                    pCurrentTri++;
                }
            }
        }

        delete [] vertMap;
    }

    bool buildStaticMesh(NMGenConfig config
        , Mesh3* sourceMesh
        , BuildContext* context
        , rcPolyMesh& polyMesh
        , rcPolyMeshDetail& detailMesh)
    {

        /**********************************************************************
         * Initialization
         *********************************************************************/

        if (!sourceMesh || sourceMesh->vertsPerPolygon != 3)
        {
            context->log(RC_LOG_ERROR, "Invalid source mesh.");
            return false;
        }

        // Some convenience variables.
        const int vertCount = sourceMesh->vertCount;
        const int triangleCount = sourceMesh->polyCount;
        const int messageDetail = context->messageDetail;

        config.applyLimits();

        // Derive various settings.

        float* bounds = new float[6];
        int height, width;

        deriveBounds3(sourceMesh->vertices
            , sourceMesh->vertCount * 3
            , bounds);
	    rcCalcGridSize(&bounds[0], &bounds[3], config.xzCellSize
            , &width, &height);

        const int vxMaxTraversableStep = 
            (int)floorf(config.maxTraversableStep / config.yCellSize);
        const int vxMinTraversableHeight =
            (int)ceilf(config.minTraversableHeight / config.yCellSize);
        const int vxTraversableAreaBorderSize = (int)ceilf(
            config.traversableAreaBorderSize / config.xzCellSize);
        const int vxHeightfieldBorderSize= (int)ceilf(
            config.heightfieldBorderSize / config.xzCellSize);
        const int vxMaxEdgeLength = (int)ceilf(
            config.maxEdgeLength / config.xzCellSize);

        // Log configuration related messages.

        if (messageDetail > 1)
        {
            context->log(RC_LOG_PROGRESS
                , "Source: %d vertices, %d triangles"
                , vertCount
                , triangleCount);
            context->log(RC_LOG_PROGRESS, "Source: %d x %d cells"
                , width, height);
        }
        if (messageDetail > 2)
        {
            context->log(RC_LOG_PROGRESS
                , "Source: Min:(%.3f, %.3f, %.3f) to Max:(%.3f, %.3f, %.3f)"
                            , bounds[0], bounds[1], bounds[2]
                            , bounds[3], bounds[4], bounds[5]);
            context->log(RC_LOG_PROGRESS
                , "Config: xzCellSize: %.3f wu"
                , config.xzCellSize);
            context->log(RC_LOG_PROGRESS
                , "Config: yCellSize: %.3f wu"
                , config.yCellSize);
             context->log(RC_LOG_PROGRESS
                , "Config: maxTraversableSlop: %.2f degrees"
                , config.maxTraversableSlope);             
            context->log(RC_LOG_PROGRESS
                , "Config: maxTraversableStep: %d vx"
                , vxMaxTraversableStep);
            context->log(RC_LOG_PROGRESS
                , "Config: minTraversableHeight: %d vx"
                , vxMinTraversableHeight);
            context->log(RC_LOG_PROGRESS
                , "Config: traversableAreaBorderSize: %d vx"
                , vxTraversableAreaBorderSize);
            context->log(RC_LOG_PROGRESS
                , "Config: heightfieldBorderSize: %d vx"
                , vxHeightfieldBorderSize);
             context->log(RC_LOG_PROGRESS
                , "Config: smoothingThreshold: %d"
                , config.smoothingThreshold);
             context->log(RC_LOG_PROGRESS
                , "Config: mergeRegionSize: %d"
                , config.mergeRegionSize);
             context->log(RC_LOG_PROGRESS
                , "Config: minIslandRegionSize: %d"
                , config.minIslandRegionSize);
            context->log(RC_LOG_PROGRESS
                , "Config: maxEdgeLength: %d vx"
                , vxMaxEdgeLength);
            context->log(RC_LOG_PROGRESS
                , "edgeMaxDeviation: %.3f wu"
                , config.edgeMaxDeviation);
            context->log(RC_LOG_PROGRESS
                , "Config: contourSampleDistance: %.3f wu"
                , config.contourSampleDistance);
            context->log(RC_LOG_PROGRESS
                , "Config: contourMaxDeviation: %.3f wu"
                , config.contourMaxDeviation);
             context->log(RC_LOG_PROGRESS
                , "Config: maxVertsPerPoly: %d"
                , config.maxVertsPerPoly);
            context->log(RC_LOG_PROGRESS
                , "Config: clipLedges: %d"
                , config.clipLedges);
        }
    	
        /**********************************************************************
         * Build solid heightfield.
         *********************************************************************/

        rcHeightfield* solidHeightfield = rcAllocHeightfield();
        bool failed = false;
	    if (!solidHeightfield)
	    {
            context->log(RC_LOG_ERROR
                , "Out of memory: Solid heightfield.");
		    failed = true;
	    }
	    else if (!rcCreateHeightfield(context
            , *solidHeightfield
            , width
            , height
            , &bounds[0]
            , &bounds[3]
            , config.xzCellSize
            , config.yCellSize))
	    {
            rcFreeHeightField(solidHeightfield);
		    context->log(RC_LOG_ERROR
                , "Could not create solid heightfield.");
		    failed = true;
	    }
    	
        delete[] bounds;
        bounds = 0;

        if (failed)
        {
            return false;
        }

        if (messageDetail > 2)
            context->log(RC_LOG_PROGRESS
                , "Initialized solid heightfield.");

	    // Allocate array that can hold triangle area information.
	    unsigned char* triangleAreas = new unsigned char[triangleCount];
	    if (!triangleAreas)
	    {
		    context->log(RC_LOG_ERROR
                , "Out of memory: triangleAreas (%d).", triangleCount);
		    return false;
	    }

    	memset(triangleAreas, 0, sizeof(unsigned char) * triangleCount);

	    // Mark triangles which are walkable based on their slope
	    rcMarkWalkableTriangles(context
            , config.maxTraversableSlope
            , sourceMesh->vertices, sourceMesh->vertCount
            , sourceMesh->indices, sourceMesh->polyCount
            , triangleAreas);

	    rcRasterizeTriangles(context
            , sourceMesh->vertices, sourceMesh->vertCount
            , sourceMesh->indices, triangleAreas, sourceMesh->polyCount
            , *solidHeightfield
            , vxMaxTraversableStep);

        delete[] triangleAreas;
        triangleAreas = 0;
    	
        if (messageDetail > 2)
            context->log(RC_LOG_PROGRESS
                , "Built solid heightfield: %d spans."
                , getSolidSpanCount(*solidHeightfield));

	    // Apply various filters.
	    rcFilterLowHangingWalkableObstacles(context
            , vxMaxTraversableStep
            , *solidHeightfield);
	    rcFilterLedgeSpans(context
            , vxMinTraversableHeight
            , vxMaxTraversableStep
            , *solidHeightfield);
	    rcFilterWalkableLowHeightSpans(context
            , vxMinTraversableHeight
            , *solidHeightfield);

        if (messageDetail > 2)
            context->log(RC_LOG_PROGRESS
                , "Applied solid heightfield filters: %d remaining spans."
                , getTraversableSpanCount(*solidHeightfield));

        /**********************************************************************
         * Build compact (open) heightfield and regions.
         *********************************************************************/

        rcCompactHeightfield* compactHeightfield = rcAllocCompactHeightfield();
	    if (!compactHeightfield)
	    {
            context->log(RC_LOG_ERROR
                , "Out of memory: Compact heightfield.");
		    failed = true;
	    }
	    else if (!rcBuildCompactHeightfield(context
            , vxMinTraversableHeight
            , vxMaxTraversableStep
            , *solidHeightfield
            , *compactHeightfield))
	    {
            rcFreeCompactHeightfield(compactHeightfield);
		    context->log(RC_LOG_ERROR
                , "Could not build compact heightfield.");
            failed = true;
	    }
    	
		rcFreeHeightField(solidHeightfield);
		solidHeightfield = 0;

        if (failed)
        {
            return false;
        }

        if (messageDetail > 2)
            context->log(RC_LOG_PROGRESS
                , "Built compact heightfield: %d of %d traversable spans."
                , getTraversableSpanCount(*compactHeightfield)
                , compactHeightfield->spanCount);

	    // Create border around traversable surface.
	    if (!rcErodeWalkableArea(context
            , vxTraversableAreaBorderSize
            , *compactHeightfield))
	    {
		    context->log(RC_LOG_ERROR
                , "Could not generate open area border.");
            rcFreeCompactHeightfield(compactHeightfield);
            return false;
	    }
    	
        if (messageDetail > 2)
            context->log(RC_LOG_PROGRESS
                , "Applied border: %d of %d tranversable spans."
                , getTraversableSpanCount(*compactHeightfield)
                , compactHeightfield->spanCount);

	    // Build distance field.
	    if (!rcBuildDistanceField(context, *compactHeightfield))
	    {
            rcFreeCompactHeightfield(compactHeightfield);
		    context->log(RC_LOG_ERROR
                , "Could not build distance field.");
		    return false;
	    }

        if (messageDetail > 2)
            context->log(RC_LOG_PROGRESS
                , "Built distance field: %d max distance."
                , compactHeightfield->maxDistance);

	    // Build regions.
	    if (!rcBuildRegions(context
            , *compactHeightfield
            , vxHeightfieldBorderSize
            , config.minIslandRegionSize
            , config.mergeRegionSize))
	    {
            rcFreeCompactHeightfield(compactHeightfield);
		    context->log(RC_LOG_ERROR
                , "Could not build regions.");
		    return false;
	    }
    	
        if (messageDetail > 2)
        {
            context->log(RC_LOG_PROGRESS
                , "Built regions: %d regions."
                , compactHeightfield->maxRegions);
            context->log(RC_LOG_PROGRESS
                , "Compact final: %d of %d traversable spans."
                , getTraversableSpanCount(*compactHeightfield)
                , compactHeightfield->spanCount);
        }

        /**********************************************************************
         * Build contours.
         *********************************************************************/

	    rcContourSet* contourSet = rcAllocContourSet();
	    if (!contourSet)
	    {
		    context->log(RC_LOG_ERROR
                , "Out of memory Contour set.");
		    failed = true;
	    }
	    else if (!rcBuildContours(context
            , *compactHeightfield
            , config.edgeMaxDeviation
            , vxMaxEdgeLength
            , *contourSet))
	    {
            rcFreeContourSet(contourSet);
		    context->log(RC_LOG_ERROR
                , "Could not create contours.");
		   failed = true;
	    }
    	
        if (failed)
        {
            rcFreeCompactHeightfield(compactHeightfield);
            return false;
        }

        if (messageDetail > 2)
            context->log(RC_LOG_PROGRESS
                , "Built contours: %d contours."
                , contourSet->nconts);

        /**********************************************************************
         * Build polygon mesh.
         *********************************************************************/

        if (!rcBuildPolyMesh(context
            , *contourSet
            , config.maxVertsPerPoly
            , polyMesh))
	    {
		    context->log(RC_LOG_ERROR
                , "Could not create polygon mesh.");
            failed = true;
	    }
    	
		rcFreeContourSet(contourSet);
		contourSet = 0;

        if (failed)
        {
            rcFreeCompactHeightfield(compactHeightfield);
            return false;
        }

        if (messageDetail > 2)
            context->log(RC_LOG_PROGRESS
                , "Built poly mesh: %d polygons."
                , polyMesh.maxpolys);

        /**********************************************************************
         * Build detail mesh.
         *********************************************************************/
    	
        if (!rcBuildPolyMeshDetail(context
            , polyMesh
            , *compactHeightfield
            , config.contourSampleDistance
            , config.contourMaxDeviation
            , detailMesh))
	    {
		    context->log(RC_LOG_ERROR
                , "Could not build detail mesh.");
            failed = true;
	    }

		rcFreeCompactHeightfield(compactHeightfield);
		compactHeightfield = 0;

        if (failed)
            return false;

        if (messageDetail > 2)
            context->log(RC_LOG_PROGRESS
                , "Built detail mesh: %d submeshes, %d vertices, %d triangles"
                , detailMesh.nmeshes
                , detailMesh.nverts
                , detailMesh.ntris);

        if (detailMesh.nverts == 0)
        {
            context->log(RC_LOG_WARNING
                , "Build process did not result in a final mesh.");
            return false;
        }

        return true;
    }
}
}
}
}

File: trunk/nav/nmgen/nmgen/nmgen/NMGen.cs
/*
 * Copyright (c) 2011 Stephen A. Pratt
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
using System;
using org.critterai.geom;
using org.critterai.nmgen.rcn;

namespace org.critterai.nmgen
{
    /// <summary>
    /// Provides various constants and utility methods related to generating
    /// navigation mesh data.
    /// </summary>
    public static class NMGen
    {
        /// <summary>
        /// The default area id used to indicate a walkable polygon.
        /// </summary>
        /// <remarks>
        /// This is the only recognized non-zero area id recognized by some 
        /// steps in the mesh build process.
        /// </remarks>
        public const byte WalkableArea = 63;

        /// <summary>
        /// The default flag applied to polygons during the build process
        /// if the <see cref="BuildFlags.ApplyPolyFlags"/> is set.
        /// </summary>
        public const ushort DefaultFlag = 0x01;

        /// <summary>
        /// Represents the null region.
        /// </summary>
        /// <remarks>
        /// <para>When a data item is given this region it is considered
        /// to have been removed from the the data set.</para>
        /// <para>Examples: When applied to a poygon, it indicates the polygon 
        /// should be culled from the final mesh. When applied to an edge,
        /// it means the edge is a solid wall.</para>
        /// </remarks>
        public const byte NullRegion = 0;

        /// <summary>
        /// Represents a null area.
        /// </summary>
        /// <remarks>
        /// <para>When a data item is given this value it is considered to 
        /// no longer be assigned to a usable area.</para>
        /// </remarks>
        public const byte NullArea = 0;

        /// <summary>
        /// The minimum allowed value for cells size parameters.
        /// </summary>
        public const float MinCellSize = 0.01f;

        /// <summary>
        /// The maximum allowed value for parameters that define maximum 
        /// vertices per polygon.
        /// </summary>
        public const int MaxAllowedVertsPerPoly = 6;

        /// <summary>
        /// The minimum value for parameters that define walkable height.
        /// </summary>
        /// <remarks>
        /// Dependencies between parameters may limit the minimum value 
        /// to a higher value.
        /// </remarks>
        public const int MinWalkableHeight = 3;

        /// <summary>
        /// The maximum allowed value for parameters that define slope.
        /// </summary>
        public const float MaxAllowedSlope = 85.0f;

        /// <summary>
        /// Returns an array with all values set to <see cref="WalkableArea"/>.
        /// </summary>
        /// <param name="size">The length of the returned array.</param>
        /// <returns>An array with all values set to <see cref="WalkableArea"/>.
        /// </returns>
        public static byte[] BuildWalkableAreaBuffer(int size)
        {
            byte[] result = new byte[size];
            for (int i = 0; i < size; i++)
            {
                result[i] = WalkableArea;
            }
            return result;
        }

        /// <summary>
        /// Set the area id of all triangles with a slope below the specified
        /// value to <see cref="WalkableArea"/>.
        /// </summary>
        /// <param name="context">The context to use duing the operation.
        /// </param>
        /// <param name="mesh">The source mesh.</param>
        /// <param name="walkableSlope">The maximum walkable slope.
        /// </param>
        /// <param name="areas">The area ids associated with each triangle.
        /// [Size: >= mesh.triCount].</param>
        /// <returns>TRUE if the operation was successful.</returns>
        public static bool MarkWalkableTriangles(BuildContext context
            , TriangleMesh mesh
            , float walkableSlope
            , byte[] areas)
        {
            if (mesh == null
                || context == null
                || areas == null || areas.Length < mesh.triCount)
            {
                return false;
            }

            NMGenEx.MarkWalkableTriangles(context.root
                , walkableSlope
                , mesh.verts
                , mesh.vertCount
                , mesh.tris
                , mesh.triCount
                , areas);

            return true;
        }

        /// <summary>
        /// Set the area id of all triangles with a slope above the specified
        /// value to <see cref="NullArea"/>.
        /// </summary>
        /// <param name="context">The context to use duing the operation.
        /// </param>
        /// <param name="mesh">The source mesh.</param>
        /// <param name="walkableSlope">The maximum walkable slope.
        /// </param>
        /// <param name="areas">The area ids associated with each triangle.
        /// [Size: >= mesh.triCount].</param>
        /// <returns>TRUE if the operation was successful.</returns>
        public static bool ClearWalkableTriangles(BuildContext context
            , TriangleMesh mesh
            , float walkableSlope
            , byte[] areas)
        {
            if (mesh == null
                || context == null
                || areas == null || areas.Length < mesh.triCount)
            {
                return false;
            }

            NMGenEx.ClearUnwalkableTriangles(context.root
                , walkableSlope
                , mesh.verts
                , mesh.vertCount
                , mesh.tris
                , mesh.triCount
                , areas);

            return true;
        }

        /// <summary>
        /// Creates <see cref="PolyMesh"/> and <see cref="PolyMeshDetail"/>
        /// data from the specified source geometry.
        /// </summary>
        /// <param name="config">The configuration to use for the build.</param>
        /// <param name="buildFlags">Flags indicating the optional build
        /// steps to peform.</param>
        /// <param name="sourceMesh">The source geometry.</param>
        /// <param name="polyMesh">The resulting polygon mesh.</param>
        /// <param name="detailMesh">The resulting detail mesh.</param>
        /// <param name="messages">The messages generated by the build.</param>
        /// <param name="trace">TRUE if detailed trace messages should
        /// be generated.</param>
        /// <returns>TRUE if the build produced valid polygon and detail 
        /// meshes.</returns>
        public static bool BuildPolyMesh(NMGenParams config
            , BuildFlags buildFlags
            , TriangleMesh sourceMesh
            , out PolyMesh polyMesh
            , out PolyMeshDetail detailMesh
            , out string[] messages
            , bool trace)
        {

            IncrementalBuilder builder = new IncrementalBuilder(trace
                , config
                , buildFlags
                , sourceMesh);

            while (!builder.IsFinished)
            {
                builder.Build();
            }

            if (builder.State == BuildState.Complete)
            {
                polyMesh = builder.PolyMesh;
                detailMesh = builder.DetailMesh;
                messages = builder.GetMessages();
                return true;
            }

            polyMesh = null;
            detailMesh = null;
            messages = builder.GetMessages();

            return false;
        }

        /// <summary>
        /// Builds an aggregate triangle mesh from a detail mesh.
        /// </summary>
        /// <remarks>
        /// <para>All duplicate vertices are merged.</para>
        /// </remarks>
        /// <param name="source">The detail mesh to extract the triangle mesh
        /// from.</param>
        /// <param name="verts">The result vertices. [Form: (x, y, z) * vertCount]
        /// </param>
        /// <param name="tris">The result triangles.
        /// [Form: (vertAIndex, vertBIndex, vertCIndex) * triCount]</param>
        /// <returns>TRUE if the operation completed successfully.
        /// </returns>
        public static bool ExtractTriMesh(PolyMeshDetail source
            , out float[] verts
            , out int[] tris)
        {
            // TODO: EVAL: Inefficient.

            verts = null;
            tris = null;
            if (source == null || source.IsDisposed || source.TriCount == 0)
                return false;

            // Assume no duplicate verts.
            float[] tverts = new float[source.VertCount * 3];
            tris = new int[source.TriCount * 3];
            int vertCount = 0;
            int triCount = 0;

            if (PolyMeshDetailEx.FlattenMesh(source
                , tverts
                , ref vertCount
                , source.VertCount
                , tris
                , ref triCount
                , source.TriCount))
            {
                verts = new float[vertCount * 3];
                Array.Copy(tverts, verts, verts.Length);
                return true;
            }

            tris = null;
            return false;
        }
    }
}
